middleware/rabbitmq/detectors-rabbitmq.tf

```diff
@@ -0,0 +1,256 @@
+    severity              = "Critical"
+    detect_label          = "CRIT"
+    disabled              = coalesce(var.sockets_disabled_critical, var.sockets_disabled, var.detectors_disabled)
+    notifications         = coalescelist(var.sockets_notifications_critical, var.sockets_notifications, var.notifications)
+    parameterized_subject = "[{{ruleSeverity}}]{{{detectorName}}} {{{readableRule}}} ({{inputs.signal.value}}) on {{{dimensions}}}"
+  }
+
+  rule {
+    description           = "is too high > ${var.sockets_threshold_warning}"
+    severity              = "Warning"
+    detect_label          = "WARN"
+    disabled              = coalesce(var.sockets_disabled_warning, var.sockets_disabled, var.detectors_disabled)
+    notifications         = coalescelist(var.sockets_notifications_warning, var.sockets_notifications, var.notifications)
+    parameterized_subject = "[{{ruleSeverity}}]{{{detectorName}}} {{{readableRule}}} ({{inputs.signal.value}}) on {{{dimensions}}}"
+  }
+}
+
+resource "signalfx_detector" "vm_memory" {
+  name = "${join("", formatlist("[%s]", var.prefixes))}[${var.environment}] RabbitMQ Node vm_memory usage"
+
+  program_text = <<-EOF
+        A = data('gauge.node.mem_used', filter=filter('plugin', 'rabbitmq') and ${module.filter-tags.filter_custom})${var.vm_memory_aggregation_function}
+        B = data('gauge.node.mem_limit', filter=filter('plugin', 'rabbitmq') and ${module.filter-tags.filter_custom})${var.vm_memory_aggregation_function}
+        signal = (A/B).scale(100).${var.vm_memory_transformation_function}(over='${var.vm_memory_transformation_window}').publish('signal')
+        detect(when(signal > ${var.vm_memory_threshold_critical})).publish('CRIT')
+        detect(when(signal > ${var.vm_memory_threshold_warning})).publish('WARN')
+  EOF
+
+  rule {
+    description           = "is approaching the watermark"
+    severity              = "Critical"
+    detect_label          = "CRIT"
+    disabled              = coalesce(var.vm_memory_disabled_critical, var.vm_memory_disabled, var.detectors_disabled)
+    notifications         = coalescelist(var.vm_memory_notifications_critical, var.vm_memory_notifications, var.notifications)
+    parameterized_subject = "[{{ruleSeverity}}]{{{detectorName}}} {{{readableRule}}} ({{inputs.signal.value}}) on {{{dimensions}}}"
+  }
+
+  rule {
+    description           = "is approaching the watermark"
+    severity              = "Warning"
+    detect_label          = "WARN"
+    disabled              = coalesce(var.vm_memory_disabled_warning, var.vm_memory_disabled, var.detectors_disabled)
+    notifications         = coalescelist(var.vm_memory_notifications_warning, var.vm_memory_notifications, var.notifications)
+    parameterized_subject = "[{{ruleSeverity}}]{{{detectorName}}} {{{readableRule}}} ({{inputs.signal.value}}) on {{{dimensions}}}"
+  }
+}
+
+resource "signalfx_detector" "messages_ready" {
+  for_each = var.messages_ready_thresholds
```

Comment:
why not using an aggregation mechanism, at least by default, to monitor every queues with the same threshold instead of having to know and configure every queues manually.

Also, I understand if this is usual to configure each queue manually with custom threshold for each one this could be painful but the "normal" way to do is :
- having a full generic detector which will works out the box
- make it possible to multiple instance the terraform module (with different filters and detectors disabled)

so here you could: 
- have a generic detector to monitor every queues by default.
- import this module once with default config and disable this detector.
- import this module for each queue you want to monitor enabling only this detector with its custom threshold.

does it make sens?

Reply:
> why not using an aggregation mechanism, at least by default, to monitor every queues with the same threshold instead of having to know and configure every queues manually.

By default it will trigger an alert for each queue, but you are right I will add an aggregation mechanism as this can lead to duplicate alert in a cluster configuration.

> does it make sens?

I thought about it and found really unpractical to call the entire module each time you want to add a custom configuration for a dedicated queue or vhost. As each detector is still created but only the alerting rules are disabled this can lead to many many unused detectors in the UI.

There are various use case where you want different thresholds and behavior and it is really unlikely that you want the same for all your queues, for example : 
* Queues with lots of messages from time to time (at night for example) where you only want to check whether your consumer are working properly : in this case you don't really care about checking if there is too many messages in the queue
* Queues where you need to consume your message really fast and you don't want too many messages in ready state : you might consider checking the ack rate and the ready message threshold

The `for_each` method might not be the right choice, maybe we can also consider to have two modules (this is what I was considering at first) : 
* `rabbitmq/node`: will contains the detectors related to the middleware itself and the various system check we need (memory, fd, sockets, process), we can add some useful cluster detectors in the future also,
* `rabbitmq/queue`: will contains the detectors related to the broker usage itself and the customer usage of it. Detectors create by this one will in most cases require some finetuning as it depends on the use case

What to you think ?

Reply:
For my part I found the for_each approach pretty convenient.

I don't understand your sentence 
> why not using an aggregation mechanism, at least by default, to monitor every queues with the same threshold instead of having to know and configure every queues manually.

currently, with messages_ack_rate_thresholds = 
```
    "default" = {
      filter             = "filter('name', '*')"
      threshold_critical = "1/60"
      threshold_warning  = "2/60"
    }
```
every queues are monitored with the same threshold (maybe I misunderstood the signalflow).

Reply:
> As each detector is still created but only the alerting rules are disabled this can lead to many many unused detectors in the UI.

It is a very good point. I confess I had in mind datadog implementation which use count on the terraform resource but here you are right.

splitting module seems to be the best approach. with terraform 0.13 we can for each on module so it will be possible to implement to loop in the "client" terraform side importing multiple times the module with different thresholds config.

from this way we can preserve the generic implementation and a straightforward usage in future.

are you ok?

Reply:
Seems to be a good compromise, just changed with the module splited in two.